### PR TITLE
Highlight all errors in VBA flow - Company step

### PR DESCRIPTION
**Pull request by aldo-expensify**
cc @marcaaron

Brought some of the functionality implemented in this [dead PR](https://github.com/Expensify/App/pull/4890)

Investigate: can functions like `clearErrorAndSetValue` be reused somehow? We will need to repeat this and other functions in every step.

### Details

Changes: 

- Added missing errors for fields in Company Step (i.e. City)
- Show all highlighted errors together in Company Step
- The "Save & continue" button is disabled if the checkbox next to "I confirm that this company...." has not been checked or if there is any field left empty. Enabling the input in all cases will be handled when we address this [GH issue](https://github.com/Expensify/App/issues/5182)

### Fixed Issues

$ https://github.com/Expensify/App/issues/4785

We are working in this issue step by step. This PR updates the error handling specifically for the company information step in the VBA flow.

### Tests AND QA Steps

1. Get to the step where we fill the "Company Information", two options:
    1. In a browser, navigate to `/bank-account/company`
    2. or do the following steps:
        1. Start with a non-public domain account that has not yet set up a workspace or bank account
        2. Create a new workspace by tapping on the green FAB button
        3. Navigate to `/bank-account` by selecting the workspace and tapping the "Get Started" button
        4. Tap button to add account manually: `Log in into your bank`
        5. Enter data following step [(3) Add Bank Account in this SO](https://stackoverflow.com/c/expensify/a/525/419)
Screen Shot 2021-09-09 at 2 42 10 PM
2. Fill all fields, and check the checkbox next to "I confirm that this company is not on the...". Make sure there is invalid data in `Zip code`, `Company website`, `Tax ID number` and `Industry classification code`. Invalid data could be for example `1`.
3. Once all fields have been filled, click "Save & continue"
4. **_Expected result_**: You should get the error modal with the generic message: `Please double check any highlighted fields and try again.` and see all fields with inline highlighted errors at the same time: Screen Shot 2021-09-13 at 5 03 35 PM
5. Focus on any of the **text** inputs with errors, type something more.
6. **_Expected result_**: the inline error and highlight should disappear only for that input. Screen Shot 2021-09-13 at 5 04 37 PM
7. Fill all fields of the form with valid values and input an invalid "Phone Number" like `123`. Valid values:
    - `Zip Code`: 123456
    - `Company website`: https://test.com
    - `Tax ID number`: 123456789
    - `Industry classification code`: 123456 
8. Click "Save & continue"
9. **_Expected result_**: We should see the loading transition (data is sent to back-end), then we should see the error modal with a message like "Please provide a valid Phone Number", and no fields should be highlighted. Screen Shot 2021-09-13 at 5 06 55 PM

### Tested On

- [x] Web
- [ ] Mobile Web
- [ ] Desktop
- [ ] iOS
- [x] Android

### CODE REVIEW

**Comment by aldo-expensify** [comment]
Update: Removed WIP/Draft. Handle error outline for StatePicker. Tested in Web and it seems to be working. I'll be testing in Android.

File: src/components/ExpensiPicker.js
Comment (Julesssss) [inline src/components/ExpensiPicker.js:45]: Is this new outer view necessary?

Comment (Julesssss) [reply src/components/ExpensiPicker.js:45]: I think it _is_ because the error text needs to live outside of the first View component. But it would be great if you could confirm that.

File: src/components/ExpensiPicker.js
Comment (Julesssss) [inline src/components/ExpensiPicker.js:64]: Unnecessary newline?

File: src/components/ExpensiPicker.js
Comment (Julesssss) [inline src/components/ExpensiPicker.js:55]: Shouldn't this Text comp be indented?

File: src/components/ExpensiTextInput/propTypes.js
Comment (Julesssss) [inline src/components/ExpensiTextInput/propTypes.js:36]: I don't see any existing usages of this prop being updated. Was this prop not in use, or do we need to update existing usages?

File: src/components/StatePicker.js
Comment (Julesssss) [inline src/components/StatePicker.js:20]: NAB: I think this comment could be slightly improved. We are not styling for errors in _this_ component, so I think it should be changed to something less specific.

File: src/pages/ReimbursementAccount/CompanyStep.js
Comment (Julesssss) [inline src/pages/ReimbursementAccount/CompanyStep.js:84]: NAB: I read this comment a few times and couldn't understand the purpose until I checked for other usages in the code.

File: src/pages/ReimbursementAccount/CompanyStep.js
Comment (Julesssss) [inline src/pages/ReimbursementAccount/CompanyStep.js:199]: NAB: This seems a bit redundant now that it's simply the inverse of `hasNoConnectionToCannabis`, but I guess it does help add some context

**Review by Julesssss — COMMENTED** [review]
I think there's some potential to improve `ExpensiTextInput` further and to delegate some of the error logic to the component, but this isn't an N6 blocker.

File: src/components/ExpensiPicker.js
Comment (aldo-expensify) [reply src/components/ExpensiPicker.js:55]: It should, thanks

File: src/components/ExpensiPicker.js
Comment (aldo-expensify) [reply src/components/ExpensiPicker.js:45]: It is because of that. If I put it inside the existing view, the error text appears inside the borders of the input.

File: src/components/ExpensiTextInput/propTypes.js
Comment (aldo-expensify) [reply src/components/ExpensiTextInput/propTypes.js:36]: I think it wasn't being used. It was declared as `hasError` in the `propTypes`, so I just corrected that misalignment  of names

File: src/components/ExpensiPicker.js
Comment (Julesssss) [inline src/components/ExpensiPicker.js:18]: NAB: It kinda feels like we're tying the prop to UI here, I would have thought the prop should simply describe state, rather than what we do with that state. But I could be wrong, and this is not a huge deal I guess.

**Review by Julesssss — COMMENTED** [review]
Tested web and the forms are working as expected. 

I see a few improvements (scroll to error on mobile, highlight API forms), but I see that we're aware of this and [still discussing here](https://expensify.slack.com/archives/C01GTK53T8Q/p1631303572177600?thread_ts=1631218542.139500&cid=C01GTK53T8Q).

**Comment by Julesssss** [comment]
Hey @marcaaron, did you want (and have time) to review this PR? If yes then I'll await your review.

File: src/components/ExpensiPicker.js
Comment (marcaaron) [reply src/components/ExpensiPicker.js:18]: What would the alternative be here? FWIW this seems to be how we're handling things in the `ExpensiTextInput`

https://github.com/Expensify/App/blob/3e110935a8cfd0fc20da4cb1560416db6baf312c/src/components/ExpensiTextInput/propTypes.js#L13-L17

File: src/components/ExpensiPicker.js
Comment (marcaaron) [inline src/components/ExpensiPicker.js:65]: Rather than use a Boolean cast can we check if the string (which defaults to `''`) is empty?

```js
!_.isEmpty(errorText) ? <Text /> : null
``` 

I think we are trying to standardize on this as per the advice [here](https://github.com/Expensify/App/blob/main/STYLE.md#important-note)

File: src/components/ExpensiTextInput/propTypes.js
Comment (marcaaron) [reply src/components/ExpensiTextInput/propTypes.js:36]: While we are here should we rename this file to `expensiTextInputPropTypes.js`? 

I noticed that the convention of using just `propTypes.js` is not correct (and all other propTypes files reference what they are propTypes for).

File: src/components/StatePicker.js
Comment (marcaaron) [reply src/components/StatePicker.js:20]: Or we could reference the `propTypes` of the `ExpensiPicker`? It looks like this is just a wrapper around this component.

File: src/pages/ReimbursementAccount/CompanyStep.js
Comment (marcaaron) [inline src/pages/ReimbursementAccount/CompanyStep.js:111]: This should be capitalized https://github.com/Expensify/App/blob/main/STYLE.md#jsdocs

File: src/pages/ReimbursementAccount/CompanyStep.js
Comment (marcaaron) [inline src/pages/ReimbursementAccount/CompanyStep.js:186]: `_.each()`

File: src/pages/ReimbursementAccount/CompanyStep.js
Comment (marcaaron) [inline src/pages/ReimbursementAccount/CompanyStep.js:187]: NAB, this works fine now, but seems potentially dangerous to assume this will always be a string e.g. if someone made a boolean a required field or changes something to an int - this would break since we can only call `.trim()` on a string.

File: src/pages/ReimbursementAccount/CompanyStep.js
Comment (marcaaron) [reply src/pages/ReimbursementAccount/CompanyStep.js:209]: FWIW we are talking about never disabling the submit button elsewhere. So this will likely be removed in a future PR. But I agree it's somewhat confusing to have the submit button _only_ disabled for this field. So I think we should either:

- Leave this logic as it was and don't allow submitting if there are missing required fields
- Remove the button disable completely

File: src/pages/ReimbursementAccount/CompanyStep.js
Comment (marcaaron) [inline src/pages/ReimbursementAccount/CompanyStep.js:250]: Are we going to add a label for this one?

File: src/pages/ReimbursementAccount/CompanyStep.js
Comment (marcaaron) [inline src/pages/ReimbursementAccount/CompanyStep.js:143]: NAB, I don't know if we need to handle this here or if we should do it in a future PR to keep things consistent for now, but wanted to call out that we are discussing [here](https://expensify.slack.com/archives/C01GTK53T8Q/p1631304509179900?thread_ts=1631218542.139500&cid=C01GTK53T8Q) that we should only clear errors on submit (possibly on blur in the future) but not on change.

**Review by marcaaron — CHANGES_REQUESTED** [review]
Looking good ! Left some comments / notes.

File: src/components/ExpensiTextInput/propTypes.js
Comment (aldo-expensify) [reply src/components/ExpensiTextInput/propTypes.js:36]: > While we are here should we rename this file to `expensiTextInputPropTypes.js`?
> 
> I noticed that the convention of using just `propTypes.js` is not correct (and all other propTypes files reference what they are propTypes for).

Ok, but I couldn't find in our style guidelines something about the naming of this file. In our codebase I can see it named starting with uppercase and starting with lowercase, which one do we prefer? I'm not sure because... I would lean toward uppercase just because components filename start with uppercase, but then, the propType is not a component itself, so it should be lowercase?

Examples:

- PickerPropTypes
- plaidLinkPropTypes
- inlineCodeBlockPropTypes
- onfidoPropTypes
- HoverablePropTypes

File: src/components/StatePicker.js
Comment (aldo-expensify) [reply src/components/StatePicker.js:20]: I guess this applies to all props in `StatePicker`, they are all passed down to `ExpensiPicker`.

Do you have any suggestion/example for a "less specific" comment or a comment referencing a `propType` of another component?

I tried to search for examples of referencing another propType while documenting in our codebase and in google and didn't find any. Would a reference to the `propTypes` of the `ExpensiPicker` look like: `See ExpensiPicker "hasError" prop documentation`?

File: src/pages/ReimbursementAccount/CompanyStep.js
Comment (aldo-expensify) [reply src/pages/ReimbursementAccount/CompanyStep.js:209]: > FWIW we are talking about never disabling the submit button elsewhere. So this will likely be removed in a future PR. But I agree it's somewhat confusing to have the submit button _only_ disabled for this field. So I think we should either:
> 
> * Leave this logic as it was and don't allow submitting if there are missing required fields
> * Remove the button disable completely

I guess I'll disable it completely. Both are a similar amount of work, and If I do this "Leave this logic as it was and don't allow submitting if there are missing required fields", we will have to work again to get to "Remove the button disable completely"

File: src/components/ExpensiPicker.js
Comment (aldo-expensify) [reply src/components/ExpensiPicker.js:18]: > NAB: It kinda feels like we're tying the prop to UI here, I would have thought the prop should simply describe state, rather than what we do with that state. But I could be wrong, and this is not a huge deal I guess.

Are you saying that the comment should be updated? or the itself prop is not well defined?

File: src/components/ExpensiPicker.js
Comment (aldo-expensify) [reply src/components/ExpensiPicker.js:65]: Sounds good, I'll try to stick to that!

File: src/components/ExpensiTextInput/propTypes.js
Comment (marcaaron) [reply src/components/ExpensiTextInput/propTypes.js:36]: I'm not sure if it's in the style guide but in my opinion 

>propType is not a component itself, so it should be lowercase 

is the correct answer. Pascal case is usually reserved for components, classes, namespaces etc.

File: src/components/StatePicker.js
Comment (marcaaron) [reply src/components/StatePicker.js:20]: To clarify, I'm suggesting that we simply import the propTypes for use here.

File: src/components/ExpensiPicker.js
Comment (marcaaron) [reply src/components/ExpensiPicker.js:65]: >I feel like it is more boilerplate for the same, not making it more readable.

When in doubt I think we should just always error on the side of consistency. If we want to change how it's done then let's change the example in the style guide then update it everywhere.

File: src/pages/ReimbursementAccount/CompanyStep.js
Comment (aldo-expensify) [reply src/pages/ReimbursementAccount/CompanyStep.js:186]: I see that `_.each` is more common than `.forEach`, and also it seems we prefer using `_.map` rather than `.map`. I understand from the point of making the code look the same to stick to what we have been doing.... but I wonder why are we sticking to use `underscore` instead of using new javascript features (not so new anymore).

Things I like about `.forEach`, `.map`, `.filter`:

- They are part of javascript, no need to use a library. From the last years developing in javascript, I have much fresher the new features from ES6 than `underscore`, which I used to use in the past.
- Mentioned array functions are easier to chain and, in my opinion, the way the expression is laid out looks more readable. (i.e. `someArray.forEach(dosomething)` looks nicer than `_.each(someArray, dosomething)`). Chaining in `underscore` has to be done by adding `_.chain` and `_.value`.

One subtle difference between the two explained [here](https://stackoverflow.com/a/18884620/16434681), but I wouldn't expect this is something we are doing here.

The advantage I can think of using `underscore` methods is that they are more ""relaxed"" and may handle more things than just arrays (i.e. you can iterate an object).

I think this last "advantage" is something that adds flexibility to our code, but it also makes it less obvious about what types we are handling. 

If I was able to choose, I would prefer to move toward relying less in `underscore` which now days is less necessary with new javascript features. Write more boiler plate when we iterate objects using code like `Object.keys(someobject).map`, and use the standard available functions in arrays like `map`, `forEach`, `filter`, etc.

Well, that's all about that subject, if we still think we are better off using `underscore` functions, I'll stick to them! 😬

File: src/pages/ReimbursementAccount/CompanyStep.js
Comment (aldo-expensify) [reply src/pages/ReimbursementAccount/CompanyStep.js:187]: Makes sense, should we add a more robust function that test if it is a string? Something like:
- if it is falsey (`undefined`, `null`, whatever is falsey), return false
- If it is a `string`, use `trim`, test with `_.isEmpty`
- If it is a `array`, test with `_.isEmpty`
- return !!`this.state[inputKey]`

A function called like `isValuePresent`

File: src/pages/ReimbursementAccount/CompanyStep.js
Comment (aldo-expensify) [reply src/pages/ReimbursementAccount/CompanyStep.js:250]: I was just putting the red outline because there is little space under the `StatePicker`, do you want me to test how it would look with the error text there?

File: src/pages/ReimbursementAccount/CompanyStep.js
Comment (aldo-expensify) [reply src/pages/ReimbursementAccount/CompanyStep.js:143]: Hmm this is how I implemented it on the first step (manually enter Routing and Account number), so it would be consistent with that.

I also think that the slack conversation is focusing on when we "validate". This code is not "validating", is just clearing the error while the user is typing. Even when the error is cleared because the user typed, it could still be "invalid" once he submits.

Comment (marcaaron) [reply src/pages/ReimbursementAccount/CompanyStep.js:143]: We can sort this out later. Mostly bringing it up since it's a design feature that we need to resolve and get feedback from others on. I agree we can leave things this way for now so that there is consistency.

File: src/pages/ReimbursementAccount/CompanyStep.js
Comment (marcaaron) [reply src/pages/ReimbursementAccount/CompanyStep.js:250]: Could we show the inline message in the space here?



Any thoughts @shawnborton ? Or should we just let this one go?

File: src/pages/ReimbursementAccount/CompanyStep.js
Comment (marcaaron) [reply src/pages/ReimbursementAccount/CompanyStep.js:187]: That might be useful. If we want to add it maybe it can go in the `ValidationUtils.js`? I could go either way here so will leave it up to you to decide if it's worth doing now or not.

File: src/components/StatePicker.js
Comment (aldo-expensify) [reply src/components/StatePicker.js:20]: I see, I just noticed that we do something like that for `BaseExpensiTextInput` and `ExpensiTextInput`. Both import the `propTypes` from the same file, and `ExpensiTextInput` is a wrapper for `BaseExpensiTextInput`.

File: src/pages/ReimbursementAccount/CompanyStep.js
Comment (aldo-expensify) [reply src/pages/ReimbursementAccount/CompanyStep.js]: Added `isValuePresent` function in `ValidationUtils.js`

**Comment by aldo-expensify** [comment]
Update: I decided to stick to leaving the "Save & continue" button disabled if there is any empty field (including terms and conditions checkbox). Why?:
- To implement the "Save & continue" always enabled behaviour I needed to add the error text and outline for another component that doesn't have it yet: `CheckboxWithLabel`.
- Adding the error outline + text to `CheckboxWithLabel` would probably require some feedback from a designer.
- I was already adding error text + outline support for `ExpensiPicker` and `StatePicker` which was feeling like going too far with this PR.

Considering this, the following changed:
- This PR is sticking to the strategy of only enabling the "Save & continue" button if there is NO empty field. This is consistent with the "add manually bank account step"
- Error text + outline for `ExpensiPicker` and `StatePicker` is not necessary in this PR anymore, so it was removed
- I'm still outlining and showing all errors together for inputs of type `ExpensiTextInput`

I'll address the following in a future PR:
- Add support for error text + outline for `ExpensiPicker`, `StatePicker` and `CheckboxWithLabel`
- Make the "Save & continue" button always enabled

File: src/pages/ReimbursementAccount/CompanyStep.js
Comment (aldo-expensify) [reply src/pages/ReimbursementAccount/CompanyStep.js]: I changed it, is it better or worse? :P

File: src/pages/ReimbursementAccount/CompanyStep.js
Comment (marcaaron) [inline src/pages/ReimbursementAccount/CompanyStep.js:84]: Ah nice our new method is useful! 😄  🎉

File: src/pages/ReimbursementAccount/CompanyStep.js
Comment (marcaaron) [inline src/pages/ReimbursementAccount/CompanyStep.js:153]: Now that we are checking if this has a value as a required field... can we remove this entirely?

File: src/pages/ReimbursementAccount/CompanyStep.js
Comment (marcaaron) [inline src/pages/ReimbursementAccount/CompanyStep.js:161]: Same for this one?

File: src/libs/ValidationUtils.js
Comment (marcaaron) [inline src/libs/ValidationUtils.js:35]: NAB, I like this method. My one hesitation is whether the boolean case makes the name confusing b.c. a boolean with a `false` is technically "present" it's just `false`. 

For the boolean case we are just returning the boolean again not whether it has a value or not.

**Review by marcaaron — CHANGES_REQUESTED** [review]
Just a couple more comments. This is looking really great.

**Comment by marcaaron** [comment]
>I was already adding error text + outline support for ExpensiPicker and StatePicker which was feeling like going too far with this PR.

Ok good call. Sounds like this is a blocker for this [issue](https://github.com/Expensify/App/issues/5182) to remove the disabled feature so we can maybe take care of this there.

>Error text + outline for ExpensiPicker and StatePicker is not necessary in this PR anymore, so it was removed

Just to clarify, since we are not able to add this for the checkbox you'd think it would be better to add inline handling later as well?

File: src/pages/ReimbursementAccount/CompanyStep.js
Comment (aldo-expensify) [reply src/pages/ReimbursementAccount/CompanyStep.js]: I decided in the end to do this: "Leave this logic as it was and don't allow submitting if there are missing required fields". There is more on this on a comment below.

File: src/libs/ValidationUtils.js
Comment (aldo-expensify) [reply src/libs/ValidationUtils.js:35]: I had the same concern about booleans, maybe it would be cleaner to consider `true` and `false` as present. Then, we just handle the case of the "terms and conditions" checkbox without this function.

File: src/pages/ReimbursementAccount/CompanyStep.js
Comment (aldo-expensify) [reply src/pages/ReimbursementAccount/CompanyStep.js:153]: You are right there.. removing..

**Comment by aldo-expensify** [comment]
> Just to clarify, since we are not able to add this for the checkbox you'd think it would be better to add inline handling later as well?

Yes. In this PR we are keeping the "Save & continue" button disabled if there is any empty field. The "Save & continue" button doesn't become enabled until we have chosen all states (StatePicker) and the company type (ExpensiPicker), so these highlights were not going to be used shown under this configuration. Same thing applies for the inline highlight of the checkbox component.

I think these things should be added together with making the "Save & continue" button always enabled, because then we will have the case where we want to inline an error in those inputs.

File: src/libs/ValidationUtils.js
Comment (aldo-expensify) [reply src/libs/ValidationUtils.js:38]: Ok, made a small change here. `false` and `true` (booleans) are considered "present". Therefore, we are not using this function to test the value for `hasNoConnectionToCannabis`.

Comment (marcaaron) [reply src/libs/ValidationUtils.js:38]: Ok cool I think this works for now. Probably we will want to think about this some more when we build a more generic form component and need a generic way to test that a field marked as "required" when a boolean will be tested for "existing" and "true".

File: src/libs/ValidationUtils.js
Comment (marcaaron) [inline src/libs/ValidationUtils.js:22]: This comment to me suggests a method that would do this kind of logic

```js
isRequirementFulfilled() {
    if (_.isString(value)) {
        return !_.isEmpty(value.trim());
    }
    if (_.isArray(value) || _.isObject(value)) {
        return !_.isEmpty(value);
    }
    if (_.isBoolean(value)) {
        return value;
    }
    return Boolean(value);
}
```

File: src/pages/ReimbursementAccount/CompanyStep.js
Comment (marcaaron) [inline src/pages/ReimbursementAccount/CompanyStep.js:137]: NAB, removing these new lines seems unnecessary. I like them personally since things are less visually compressed, but that's just my opinion and there are no style rules about this as far as I know.

**Review by marcaaron — CHANGES_REQUESTED** [review]
LGTM. I only really have one small note left about the comment in the `isValuePresent` method. Once that is resolved I'm ready to merge.

File: src/pages/ReimbursementAccount/CompanyStep.js
Comment (aldo-expensify) [reply src/pages/ReimbursementAccount/CompanyStep.js:137]: Got mixed up with the [php guidelines](https://github.com/Expensify/Style-Guides/blob/master/php.md#blank-lines). I assumed that a similar rule would apply here... reverting!

File: src/libs/ValidationUtils.js
Comment (aldo-expensify) [reply src/libs/ValidationUtils.js:22]: Hmm, that would be assuming `false` is not valid value for a "required" field, right?

I changed this because it is not so obvious that `false` means that the value is missing. You could a field with possible values `false` | `true` | `undefined`, for example, and interface with two radio buttons "Yes" (`true`) and "No" (`false`) and none of them start selected (`undefined`). In this case, `true` and `false` would suffice if you are just validating "required". But then, this is not really a case we have right now, so an option is to stick to the previous way of the function and think about it if we get to the case where this can be a problem:

```
isRequirementFulfilled() {
    if (_.isString(value)) {
        return !_.isEmpty(value.trim());
    }
    if (_.isArray(value) || _.isObject(value)) {
        return !_.isEmpty(value);
    }
    return Boolean(value);
}
```

About the name of the function `isRequirementFulfilled`, I feel like `requirement` is broad and could be any kind of requirement, not a value being "required" (or mandatory), but then it could be because I'm not a native english speaker :P. Meanwhile, `isRequiredFulfilled` may be better because it is about the "required" requirement.

File: src/libs/ValidationUtils.js
Comment (aldo-expensify) [reply src/libs/ValidationUtils.js:38]: > Ok cool I think this works for now. Probably we will want to think about this some more when we build a more generic form component and need a generic way to test that a field marked as "required" when a boolean will be tested for "existing" and "true".

Yes, I think this is a good idea. Postpone thinking further about this until we have more cases or a more generic.

File: src/libs/ValidationUtils.js
Comment (marcaaron) [reply src/libs/ValidationUtils.js:22]: >Hmm, that would be assuming false is not valid value for a "required" field, right?

All of the scenarios we have today suggest a checkbox that is not checked (`false`) should not pass a "required" field constraint. 

>You could a field with possible values false | true | undefined, for example, and interface with two radio buttons

I understand your example, but I would maybe ask that if we are requiring a response to a true/false question then why are we giving the option to input `undefined` at all? Will we ever have a need for something like "boolean, but 'required' means 'true' or 'false', but not 'undefined'"?

>I feel like requirement is broad and could be any kind of requirement, not a value being "required" (or mandatory), but then it could be because I'm not a native english speaker

Makes sense to me! I like the suggestion to make this `isRequiredFulfilled`.

Comment (aldo-expensify) [reply src/libs/ValidationUtils.js:22]: > > Hmm, that would be assuming false is not valid value for a "required" field, right?
> 
> All of the scenarios we have today suggest a checkbox that is not checked (`false`) should not pass a "required" field constraint.
> 
> > You could a field with possible values false | true | undefined, for example, and interface with two radio buttons
> 
> I understand your example, but I would maybe ask that if we are requiring a response to a true/false question then why are we giving the option to input `undefined` at all? Will we ever have a need for something like "boolean, but 'required' means 'true' or 'false', but not 'undefined'"?

I think you are pointing to that I'm overthinking/over-engineering this. If it is that, I agree, because we don't have such cases now and maybe we won't have them. If we stick to what we are dealing with right now, as you say, the `false` value should not pass the "required" field constraint.

If you agree with that, I'll revert to how it was working at first.

Comment (marcaaron) [reply src/libs/ValidationUtils.js:22]: >If you agree with that, I'll revert to how it was working at first. This would also include putting hasNoConnectionToCannabis back in this.requiredFields.

Yep let's do it and get this merged.

**Comment by aldo-expensify** [comment]
Update:
- `isValuePresent` became `isRequiredFulfilled`
- `isRequiredFulfilled` returns `false` if `value` is `false`
- Added `hasNoConnectionToCannabis` to required fields in `CompanyStep.js`

**Review by marcaaron — APPROVED** [review]

**Comment by marcaaron** [comment]
Going to merge this so we can keep plugging away at the remaining steps. @Julesssss thanks for your review and please let us know if you've got any additional comments that we can follow up on if they were not resolved.

**Pull request merged**
